### PR TITLE
Use a custom hashing function for remotecall_fetch

### PR DESCRIPTION
**Pull request by mbauman**
Mitigates issue JuliaLang/Distributed.jl#48 (but does not fix it).

This implements a custom hashing function for Distributed to use in determining if it should resend values between workers. Previously, this just used `Base.hash`. This has the possibility that a custom type could opt for a cheaper "partial" or constant-valued hash function, leading to failure to update values between workers. Instead, this implements a generic hashing scheme that works directly on the internal fields --- we don't need to worry about hash equality between different types for this use-case.

### CODE REVIEW

File: stdlib/Distributed/src/clusterserialize.jl [hunk 1]
```diff
@@ -44,6 +44,31 @@ end
...
+end
+hashall(x::Type, h::UInt) = hash(x, h)
+hashall(x::Module, h::UInt) = hash(x, h)
+function hashall(A::Array, h::UInt)
+    for i in eachindex(A)
+        h = hashall(isassigned(A, i) ? A[i] : 0x0a57083376608372, h)
+    end
+    return h
+end
+function hashall(S::String, h::UInt)
```
Comment (JeffBezanson) [inline stdlib/Distributed/src/clusterserialize.jl]: This can just call `hash`.

**Comment by raminammour** [comment]
First, thank you for working on this!

Will this be merged soon? 

Note that the behavior on `v1.0` is:
```
julia> @everywhere begin
       mutable struct S4
           val
           end
       end

julia> x=S4(1)
S4(1)

julia> remotecall_fetch(()->x, 2)
S4(1)

julia> x.val=16
16

julia> remotecall_fetch(()->x, 2)
S4(1)
```
Which may surprise users, I was certainly surprised the first time it arose in an application. 

Maybe clarifying the following sentence in the docs:

`Globals are re-sent to a destination worker only in the context of a remote call, and then only if its value has changed*`

*by noting that the the "change" is detected by `hash(x)`, and directing the user to implement a hash function for their type would be helpful given the current behavior.

**Comment by mbauman** [comment]
Sure, let's try to get this in soon.  I'll close and re-open to re-trigger CI.

**Comment by raminammour** [comment]
This was never merged, is there anything blocking it? 

Also, now it appears that this code has a bug regardless of hash collisions (see #31252).

**Comment by mbauman** [comment]
No blockers on my side, but I would appreciate a review from someone more familiar with the remotecall internals.

File: stdlib/Distributed/src/clusterserialize.jl [hunk 2]
```diff
@@ -44,6 +44,26 @@ end
 
 const known_object_data = Dict{UInt64,Any}()
 
+# Make issue #25220 less likely by ensuring we hash all internal values
+hashall(x) = hashall(x, UInt(0))
+hashall(x, h::UInt) = fieldcount(typeof(x)) == 0 ? hash(x, h) : hashfields(x, h)
```
Comment (JeffBezanson) [inline stdlib/Distributed/src/clusterserialize.jl]: Can use `nfields(x)` instead.

File: stdlib/Distributed/src/clusterserialize.jl [hunk 3]
```diff
@@ -44,6 +44,26 @@ end
...
+@noinline function hashfields(@nospecialize(x), h::UInt)
+    for i in 1:fieldcount(typeof(x))
+        h = hashall(isdefined(x, i) ? getfield(x, i) : 0x00e1dfd11f636805, h)
```
Comment (JeffBezanson) [inline stdlib/Distributed/src/clusterserialize.jl]: The constant should be truncated to `UInt` (for 32-bit).

**Comment by vtjnash** [comment]
Moved to https://github.com/JuliaLang/Distributed.jl/issues/50

**Pull request closed without merging**